workflow/controller/controller.go

```diff
@@ -523,32 +531,49 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
 	logCtx.Info("cleaning up pod")
 	err := func() error {
 		pods := wfc.kubeclientset.CoreV1().Pods(namespace)
+		pod, err := pods.Get(ctx, podName, metav1.GetOptions{})
```

Comment:
The SharedInformer pod information is sometimes out of date(see: [191/files#r1399927221](https://github.com/rohankmr414/argo-workflows/pull/191/files#r1399927221)) && pods already have a pod information from the API, so I use it instead of the `getPod` function.

Reply:
So, we'll now have 3 or so more API calls per pod I guess? Would you mind summarizing why this doesn't simply cause the informer to be on a delay and for everything to just happen on a delay? (which would seem acceptable)

Reply:
The informer can certainly be out of date. However, 3 APIs calls per pod seems high. Could you have an idempotent patch

Reply:
Sorry, I edited my original comment above, which didn't really make sense, and took a better look at the link you posted in the top of this thread @sakai-ast (apologies that I should have looked at that more closely before). Trying to understand your comment @alexec. Which Patch Operation do you consider not to be idempotent? There's a setting of a label to True, and there's a removal of a Finalizer at a given index. It seems like we could do each of them multiple times; it's just that the latter may give an error if it doesn't exist, but to me that theoretically seems like it should be okay (except that an error message could be misleading).

Reply:
Thinking about the API calls again. If we decide we still need them, then maybe we only need them for the calls to `labelPodCompleted` and `deletePod`?

Reply:
I think so.

Reply:
I have restored the `terminateContainers`, `killContainers`, and `signalContainers()` logic, and modified it so that the logic for obtaining `pods` and `pod` information from the API is only called by `labelPodCompleted` and `deletePod`. 
This adjustment should lead to a more appropriate number of API calls.
https://github.com/argoproj/argo-workflows/pull/12413/commits/96358f98ec9736f417499c50dd1a3c54b519e8bc